extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala

```diff
@@ -192,26 +192,55 @@ case class FinalStageResourceManager(session: SparkSession)
       return
     }
 
-    // Note, `SparkContext#killExecutors` does not allow with DRA enabled,
-    // see `https://github.com/apache/spark/pull/20604`.
-    // It may cause the status in `ExecutorAllocationManager` inconsistent with
-    // `CoarseGrainedSchedulerBackend` for a while. But it should be synchronous finally.
-    //
-    // We should adjust target num executors, otherwise `YarnAllocator` might re-request original
-    // target executors if DRA has not updated target executors yet.
-    // Note, DRA would re-adjust executors if there are more tasks to be executed, so we are safe.
-    //
-    //  * We kill executor
-    //      * YarnAllocator re-request target executors
-    //         * DRA can not release executors since they are new added
-    // ----------------------------------------------------------------> timeline
+    val startTime = System.currentTimeMillis()
+    val deadline = startTime +
+      conf.getConf(KyuubiSQLConf.FINAL_WRITE_STAGE_MAX_WAIT_TIME_BEFORE_KILL)
+    var draTargetExecutors =
+      getDraTargetExecutors(sc, executorAllocationClient).getOrElse(0)
+    while (System.currentTimeMillis() < deadline && draTargetExecutors > targetExecutors) {
+        Thread.sleep(100)
+        draTargetExecutors =
+            getDraTargetExecutors(sc, executorAllocationClient).getOrElse(0)
+    }
+    logInfo(s"DRA target executor number changed to $draTargetExecutors in " +
+      s"${System.currentTimeMillis() - startTime} ms.")
+
+    if (draTargetExecutors <= targetExecutors) {
+      // Ensure target executor number has been updated in cluster manager client
+      executorAllocationClient.requestExecutors(0)
```

Comment:
I'm not sure I get it correctly. It seems we should do nothing if the dra target executor is smaller than our target executor. If it happens, the executors are pending to kill.